modules/adapters/slurm-collector/src/main.rs

```diff
@@ -0,0 +1,382 @@
+    "JobID",
+    "Account",
+    "Start",
+    "ElapsedRaw",
+    // optional
+    "End",
+    "State",
+    "ReqCPUS",
+    "ReqMem",
+    "Partition",
+];
+
+impl SacctExecutor for SacctExecutorImpl {
+    fn sacct(&self, since: DateTime<Utc>) -> Result<BoxStream<Result<SacctData>>> {
+        let since_str = since.format(SACCT_DATETIME_FORMAT).to_string();
+
+        log::info!("Fetching sacct data since {}", since_str);
+
+        // with --state we need an endtime
+        let end_time_str = Utc::now().format(SACCT_DATETIME_FORMAT).to_string();
+
+        let child = Command::new("sacct")
+            .args([
+                "-n", // no header
+                "-a", // all users
+                "-P", // parseable, adds "|" delimiter
+                "-X", //remove step details (JobID.batch , .ext etc)
+                "--format",
+                SACCT_FIELDS.join(",").as_str(),
+                "--state",
+                "COMPLETED,FAILED,TIMEOUT,PREEMPTED,OUT_OF_MEMORY,CANCELLED", // TODO make configurable https://slurm.schedmd.com/sacct.html#SECTION_JOB-STATE-CODES
+                "--starttime",
+                &since_str,
+                "--endtime",
+                &end_time_str,
+            ])
+            .stdout(Stdio::piped())
+            .spawn()
+            .expect("Failed to spawn sacct command");
+
+        let stdout = child.stdout.ok_or(AppError::SacctError(
+            "Failed to open sacct stdout".to_string(),
+        ))?;
+        let reader = BufReader::new(stdout);
+
+        Ok(futures::stream::iter(reader.lines().map(move |line| {
+            let line = line.map_err(AppError::IoError)?;
+            parse_sacct_line(&line)
+        }))
+        .boxed())
+    }
+}
+
+async fn process_sacct_data<T: SacctExecutor>(
+    config: &AppConfig,
+    client: &mut GrpcClient,
+    sacct_executor: &T,
+) -> Result<()> {
+    let mut checkpoint = load_checkpoint(&config.state_file, &config.initial_checkpoint)?;
+
+    let mut sacct_data_stream = sacct_executor.sacct(checkpoint.last_processed_time)?;
+    let mut batch = Vec::with_capacity(config.batch_size);
+
+    while let Some(data) = sacct_data_stream.next().await {
+        let data = data?;
+        // Skip already processed jobs for the current timestamp
+        if checkpoint.last_processed_time == data.start_time
+            && checkpoint.processed_jobs.contains(&data.job_id)
+        {
+            continue;
+        }
+
+        batch.push(data.clone());
+
+        if batch.len() >= config.batch_size {
+            send_batch_to_api(client, &batch).await?;
+            update_checkpoint(&mut checkpoint, &batch);
+            save_checkpoint(&config.state_file, &checkpoint)?;
+
+            batch.clear();
+        }
+    }
+
+    // Process any remaining data
+    if !batch.is_empty() {
+        send_batch_to_api(client, &batch).await?;
+        update_checkpoint(&mut checkpoint, &batch);
+        save_checkpoint(&config.state_file, &checkpoint)?;
```

Comment:
I would merge those 2 function in a single one, Imo it would be more readable:
 - we update the checkpoint only when the last element is not None
 - but we save it in all cases, even if it wasn't updated at all
 
I understand that in our case batch won't be empty when we invoke update_checkpoint so the checkpoint will be updated in all cases but this is not so obvious (at least to me). Moving the write to file part into the update_checkpoint might improve the readability and also maintainability.